File: Sources/ExcalidrawModel/SceneDocument.swift
import Foundation

/// Serialization between a `Scene` and `.excalidraw` document bytes. Loading
/// always passes through `Restore` so older/partial files are canonicalised.
/// This is the persistence core the document browser / autosave build on.
public enum SceneDocument {
    public static let fileExtension = "excalidraw"
    public static let utType = "com.excalidraw.scene"

    public static func encode(
        _ scene: Scene, source: String = "excalidraw-swift", prettyPrinted: Bool = true
    ) throws -> Data {
        try scene.toFile(source: source).jsonData(prettyPrinted: prettyPrinted)
    }

    public static func decode(_ data: Data) throws -> Scene {
        try Scene(file: Restore.restore(ExcalidrawFile.decode(from: data)))
    }
}

File: web/packages/excalidraw-svelte/src/model/scene-document.ts
import { decodeFile, encodeFile } from "./file.js";
import { restore } from "./restore.js";
import { Scene } from "./scene.js";

/**
 * Serialization between a `Scene` and `.excalidraw` document text. Loading
 * always passes through `restore` so older/partial files are canonicalised.
 * (parity: SceneDocument.swift)
 */
export const SceneDocument = {
  fileExtension: "excalidraw",

  encode(scene: Scene, source = "excalidraw-web", prettyPrinted = true): string {
    return encodeFile(scene.toFile(source), prettyPrinted);
  },

  decode(json: string): Scene {
    return Scene.fromFile(restore(decodeFile(json)));
  },
} as const;
